cg/meta/workflow/mutant/metadata_parser/utils.py

```diff
@@ -0,0 +1,52 @@
+from cg.constants.constants import ControlOptions
+from cg.exc import LimsDataError
+from cg.store.models import Case, Sample
+from cg.apps.lims.api import LimsAPI
+from cg.constants.lims import LimsArtifactTypes, LimsProcess
+from genologics.entities import Artifact
+from genologics.entities import Sample as LimsSample
+from cg.models.cg_config import LOG
+
+
+def is_sample_external_negative_control(sample: Sample) -> bool:
+    return sample.control == ControlOptions.NEGATIVE
+
+
+def get_negative_controls_from_list(samples: list[LimsSample]) -> list[LimsSample]:
+    """Filter and return a list of internal negative controls from a given sample list."""
+    negative_controls = []
+    for sample in samples:
+        if sample.udf.get("Control") == "negative" and sample.udf.get("customer") == "cust000":
+            negative_controls.append(sample)
+    return negative_controls
+
+
+def get_internal_negative_control_id_from_lims(lims: LimsAPI, sample_internal_id: str) -> str:
+    """Retrieve from lims the sample_id for the internal negative control sample present in the same pool as the given sample."""
+    try:
+        artifact: Artifact = lims.get_latest_artifact_for_sample(
+            LimsProcess.COVID_POOLING_STEP, LimsArtifactTypes.ANALYTE, sample_internal_id
+        )
+        samples = artifact[0].samples
+
+        negative_controls: list = get_negative_controls_from_list(samples=samples)
+
+        if len(negative_controls) > 1:
+            sample_ids = [sample.id for sample in negative_controls]
+            LOG.warning(f"Several internal negative control samples found: {' '.join(sample_ids)}")
+        else:
+            return negative_controls[0].id
+    except Exception as exception_object:
+        raise LimsDataError from exception_object
+
+
+def get_internal_negative_control_id(lims: LimsAPI, case: Case) -> str:
+    """Query lims to retrive internal_negative_control_id."""
+
+    sample_internal_id = case.sample_ids[0]
+
+    internal_negative_control_id: str = get_internal_negative_control_id_from_lims(
+        lims=lims, sample_internal_id=sample_internal_id
+    )
+
+    return internal_negative_control_id
```

Comment:
```suggestion
def get_internal_negative_control_id(lims: LimsAPI, sample_internal_id: str) -> str:
    """Query lims to retrive internal_negative_control_id."""

    internal_negative_control_id: str = get_internal_negative_control_id_from_lims(
        lims=lims, sample_internal_id=sample_internal_id
    )

    return internal_negative_control_id
```

Reply:
So I definitely agree that it is very strange to have a hardcoded `case.sample_ids[0]` but unfortunately there is no option to get a lims pool artifact for a case, only for a specific sample. For mutant samples, all samples in the case are sequenced in the same pool so here `case.sample_ids[0]` is just one sample from the case used in order to reach the lims pool artifact.

I don't know how to improve this so if you have any suggestions, do please let me know!

Reply:
My impression is that @ChrOertlin just did not think the case should be the input. So basically, get the sample_internal_id before calling this method and then pass only that. Makes it a bit more flexible if you want a specific sample in the future which does not necessarily show as the first in the list.

Reply:
I understand that, but what I am looking for here is to have a function that receives a case and retrieves the control for that case. So it doesn't make sense to get the sample_internal_id before calling it. 🤔

Reply:
Will resolve this conversation and tag you in the corresponding function in the updated code!